cryptography/src/handshake.rs

```diff
@@ -0,0 +1,246 @@
+use chacha20poly1305::{aead::AeadMut, ChaCha20Poly1305, KeyInit};
+use commonware_codec::Encode;
+use rand_core::CryptoRngCore;
+
+use crate::{
+    transcript::{Summary, Transcript},
+    PublicKey, Signer, Verifier,
+};
+
+const NAMESPACE: &'static [u8] = b"commonware/handshake";
+
+pub struct Msg1<S> {
+    time_ms: u64,
+    epk: x25519_dalek::PublicKey,
+    sig: S,
+}
+
+pub struct Msg2<S> {
+    time_ms: u64,
+    epk: x25519_dalek::PublicKey,
+    sig: S,
+    confirmation: Summary,
+}
+
+pub struct Msg3 {
+    confirmation: Summary,
+}
+
+pub struct DialState<P> {
+    esk: x25519_dalek::EphemeralSecret,
+    peer_identity: P,
+    transcript: Transcript,
+}
+
+pub struct ListenState {
+    confirmation: Summary,
+    send: SendCipher,
+    recv: RecvCipher,
+}
+
+struct CounterNonce {
+    inner: u128,
+}
+
+impl CounterNonce {
+    pub fn new() -> Self {
+        Self { inner: 0 }
+    }
+
+    pub fn inc(&mut self) -> [u8; 16] {
+        if self.inner >= 1 << 96 {
```

Comment:
Alternatively, we could return option here I guess for the stream to restart?

Reply:
Yeah, well aware of the issue with reuse (moreso noting this is an API change where we'd otherwise error rather than panic...regardless of how likely that is 😓 )

Reply:
Yeah, if you sent 1 million messages a second, it would be 2 billion years or so to exhaust the counter. I think for galactic scale improbabilities like this it's best to not present a Result to the user of the API, but if that's idiomatic, that would work as well.

Reply:
Ended up creating an error message for this case